### PR TITLE
feat(API): list groups that can be deleted

### LINKED ISSUE #2246: Check if user has a deletable group via REST API
### Description
There's a need for the API to check if a particular user has any group that can be deleted.

#### Steps to be followed

1. Create new route in the main page (index.php) as a part of Group routes.
2. Add a new function in GroupController to check for the group's existence.
3. Test the API.
4. Add the new API's documentation  in [https://github.com/fossology/fossology/blob/master/src/www/ui/api/documentation/openapi.yaml](url)
5. Open a pull request.

#### Expected Results

1.  An endpoint to be accessed at ''/groups/has-deletable-group".
2. The API should return the object with the property **available** marked as 
   **true** if the API exists else as **false**.

cc: @shaheemazmalmmd @GMishx

### PR DESCRIPTION
**Pull request by dushimsam**
Signed-off-by: dushimsam <dushsam100@gmail.com>

## Description

To check if the user has any group that is possible to be deleted.

### Changes

1. Added a new function in `GroupController`. to check if the group exists.
2. Updated  the main file(`index.php`) by adding a new route `/groups/has-deletable-group`.
3. Updated the `openapi.yaml` file to introduce a new API.

## How to test

1. Open and run the project then visit  the url `http://localhost/repo/api/v1/groups/has-deletable-group`
2. Response

The following property in the response body called `available` should be marked as **true** if you have any group that is deletable  , else it will always be marked as **false**.

has-deletable-group

By default you have group `fossy` , this one is not deletable , the deletable groups are the one you add afterwards.
In other words , by default you don't have any deletable group , unless you start adding your own groups.

### Related Issue:
Fixes #2246
    
cc: @shaheemazmalmmd @GMishx

<a href="https://gitpod.io/#https://github.com/fossology/fossology/pull/2247"></a>

### CODE REVIEW

File: src/www/ui/api/Controllers/GroupController.php [hunk 1]
```diff
@@ -92,4 +93,28 @@ public function createGroup($request, $response, $args)
...
+   * Check if user has a group that can be deleted
+   *
+   * @param ServerRequestInterface $request
+   * @param ResponseHelper $response
+   * @param array $args
+   * @return ResponseHelper
+   */
+  public function hasDeletableGroup($request, $response, $args)
+  {
+    $userId = Auth::getUserId();
```
Comment (GMishx) [inline src/www/ui/api/Controllers/GroupController.php:107]:
```suggestion
    $userId = $this->restHelper->getUserId();
```

File: src/www/ui/api/documentation/openapi.yaml [hunk 2]
```diff
@@ -1144,7 +1144,23 @@ paths:
                 $ref: '#/components/schemas/Info'
         default:
             $ref: '#/components/responses/defaultResponse'
-
+  /groups/has-deletable-group:
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:1147]: Please check the comment on issue #2246 as a suggestion for the name change.

File: src/www/ui/api/documentation/openapi.yaml [hunk 3]
```diff
@@ -2156,6 +2172,12 @@ components:
               enum:
                 - OK
                 - ERROR
+    GroupAvailable:
+      type: object
+      properties:
+        available:
+          type: boolean
+          description: deletable group availability identifier
```
Comment (GMishx) [inline src/www/ui/api/documentation/openapi.yaml:2180]: IMO a more useful response will be a list of groups which the current user can delete (as it is returned by the `getDeletableAdminGroupMap()`).

By doing so, the UI can quickly remove groups which can not be deleted from a potential drop-down saving bunch of failing REST API calls.

As a suggestion, the response can look something like the following when the user can delete something:
```json
{
  "groups": [
      3,
      4,
      ...
    ]
}
```

Or like following when they can not delete any groups:
```json
{
  "groups": []
}
```

**Review by GMishx — CHANGES_REQUESTED** [review]
Some suggestions

File: src/www/ui/api/documentation/openapi.yaml (hunk 3 above)
Comment (dushimsam) [reply src/www/ui/api/documentation/openapi.yaml:2180]: Noted

**Comment by GMishx** [comment]
@dushimsam , can you please fix the branch so we can merge it?

**Comment by dushimsam** [comment]
Done, you can review and let me know what's next to update if still available.

**Comment by dushimsam** [comment]
Completed tests for this controller. 
Hope it can be merged now.

**Review by GMishx — APPROVED** [review]
Changes looks good.

**Review by GMishx — APPROVED** [review]
Changes looks good. Working as expected.

File: src/www/ui_tests/api/Controllers/GroupControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+   * -# Test GroupController::getDeletableGroups()
+   * -# Check if the response is a list of groups.
+   */
+  public function testGetDeletableGroups()
+  {
+    $userId = 2;
+    $groupList = array();
+    $this->restHelper->shouldReceive('getUserId')->andReturn($userId);
+    $this->userDao->shouldReceive('getAdminGroupMap')->andReturn([]);
+    $expectedResponse = (new ResponseHelper())->withJson($groupList, 200);
```
Comment (GMishx) [inline src/www/ui_tests/api/Controllers/GroupControllerTest.php:147]: Looks like you forgot to import the `ResponseHelper` class.

**Comment by dushimsam** [comment]
tested and working perfectly. 
Please merge for me,

File: src/www/ui_tests/api/Controllers/GroupControllerTest.php [hunk 5]
```diff
@@ -0,0 +1,158 @@
+<?php
+/***************************************************************
+ * Copyright (C) 2020 Siemens AG
```
Comment (GMishx) [inline src/www/ui_tests/api/Controllers/GroupControllerTest.php:3]: Please remove Siemens' copyright and add yours 😃

**Review by GMishx — CHANGES_REQUESTED** [review]
A tiny change. Missed it earlier.

File: src/www/ui_tests/api/Controllers/GroupControllerTest.php (hunk 5 above)
Comment (dushimsam) [reply src/www/ui_tests/api/Controllers/GroupControllerTest.php:3]: alright thanks

Comment (dushimsam) [reply src/www/ui_tests/api/Controllers/GroupControllerTest.php]: done

File: src/www/ui_tests/api/Controllers/GroupControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+ * version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
```
Comment (GMishx) [inline src/www/ui_tests/api/Controllers/GroupControllerTest.php:16]: As you are aware, FOSSology is now REUSE compliant, kindly change the comment like following
```suggestion
 * SPDX-FileCopyrightText: © 2022 Samuel Dushimimana <dushsam100@gmail.com>
 *
 * SPDX-License-Identifier: GPL-2.0-only
```

**Review by GMishx — CHANGES_REQUESTED** [review]
@dushimsam Kindly rebase the branch with upstream master to resolve the merge conflicts.

**Comment by dushimsam** [comment]
Ready now sir, kindly review and merge for me.

File: src/www/ui_tests/api/Controllers/GroupControllerTest.php [hunk 7]
```diff
@@ -107,35 +108,40 @@ private function getResponseJson($response)
...
-  public function testDeleteGroup()
+  public function testGetDeletableGroups()
   {
-    $groupId = 4;
-    $userId = 1;
+    $userId = 2;
+    $groupList = array();
     $this->restHelper->shouldReceive('getUserId')->andReturn($userId);
-    $this->dbHelper->shouldReceive('doesIdExist')
-      ->withArgs(["groups", "group_pk", $groupId])->andReturn(true);
```
Comment (GMishx) [inline src/www/ui_tests/api/Controllers/GroupControllerTest.php:123]: @dushimsam Looks like during rebase, you removed `testDeleteGroup()` :-|

Comment (dushimsam) [reply src/www/ui_tests/api/Controllers/GroupControllerTest.php]: Oh thanks

Comment (dushimsam) [reply src/www/ui_tests/api/Controllers/GroupControllerTest.php]: hope things are okay now.

**Review by GMishx — APPROVED** [review]
